src/main/scala/com/fulcrumgenomics/bam/SamRecordClipper.scala

```diff
@@ -0,0 +1,303 @@
+    * @param numberOfBasesToClip the number of additional bases to clip beyond any clipping
+    *                            already applied to the read
+    * @param mode the kind of clipping to be performed
+    */
+  def clipEndOfAlignment(rec: SAMRecord, numberOfBasesToClip: Int, mode: ClippingMode) : Unit = {
+    if (rec.getReadUnmappedFlag || numberOfBasesToClip < 1) {
+      Unit
+    }
+    else if (numberOfClippableBases(rec) <= numberOfBasesToClip) {
+      SAMUtils.makeReadUnmapped(rec)
+    }
+    else {
+      val oldElems = rec.getCigar.toSeq.reverse
+      val oldBases = rec.getReadBases.reverse
+      val oldQuals = rec.getBaseQualities.reverse
+      val (newElems, readBasesClipped, refBasesClipped, bases, quals) = clip(oldElems, numberOfBasesToClip, oldBases, oldQuals, mode)
+      rec.setCigar(new Cigar(util.Arrays.asList(newElems.reverse:_*)))
+      rec.setReadBases(bases.reverse)
+      rec.setBaseQualities(quals.reverse)
+    }
+  }
+
+  /**
+    * Attempts to clip an additional numberOfBasesToClip from the 5' end of the read. For
+    * details see [[clipStartOfAlignment()]] and [[clipEndOfAlignment()]].
+    *
+    * @param rec the record to be clipped
+    * @param numberOfBasesToClip the number of additional bases to be clipped
+    * @param mode the clipping mode
+    */
+  def clip5PrimeEndOfAlignment(rec: SAMRecord, numberOfBasesToClip: Int, mode: ClippingMode) = {
+    if (!rec.getReadUnmappedFlag && rec.getReadNegativeStrandFlag) clipEndOfAlignment(rec, numberOfBasesToClip, mode)
+    else clipStartOfAlignment(rec, numberOfBasesToClip, mode)
+  }
+
+  /**
+    * Attempts to clip an additional numberOfBasesToClip from the 5' end of the read. For
+    * details see [[clipStartOfAlignment()]] and [[clipEndOfAlignment()]].
+    *
+    * @param rec the record to be clipped
+    * @param numberOfBasesToClip the number of additional bases to be clipped
+    * @param mode the clipping mode
+    */
+  def clip3PrimeEndOfAlignment(rec: SAMRecord, numberOfBasesToClip: Int, mode: ClippingMode) = {
+    if (!rec.getReadUnmappedFlag && rec.getReadNegativeStrandFlag) clipStartOfAlignment(rec, numberOfBasesToClip, mode)
+    else clipEndOfAlignment(rec, numberOfBasesToClip, mode)
+  }
+
+  /**
+    * Ensures that there are at least clipLength bases clipped at the start (left-hand end)
+    * of the read, _including_ any existing soft and hard clipping.  Calculates any
+    * additional clipping and delegates to [[clipStartOfAlignment()]].
+    *
+    * @param rec the record to be clipped
+    * @param clipLength the total amount of clipping desired, including any existing clipping
+    * @param mode the clipping mode
+    */
+  def clipStartOfRead(rec: SAMRecord, clipLength: Int, mode: ClippingMode) = {
+    val existingClipping = rec.getCigar.getCigarElements.takeWhile(_.getOperator.isClipping).map(_.getLength).sum
+    if (clipLength > existingClipping) clipStartOfAlignment(rec, clipLength - existingClipping, mode)
```

Comment:
Let's chat about this one.  I sort of agree, but don't really want to take this on right now.